neal/sampler.py

```diff
@@ -239,21 +269,62 @@ def sample(self, bqm, beta_range=None, num_reads=None, num_sweeps=1000,
+            if beta_schedule is None:
+                error_msg = "'beta_schedule' must be provided for beta_schedule_type = 'custom': value is None"
+                raise ValueError(error_msg)
+            elif num_sweeps is not None and num_sweeps!=len(beta_schedule)*num_sweeps_per_beta:
+                error_msg = "'num_sweeps' should be set to None, or a value consistent with 'beta_schedule' and 'num_sweeps_per_beta' for 'beta_schedule_type' = 'custom': value = " + str(num_sweeps)
+                raise ValueError(error_msg)
+            elif beta_range is not None and (beta_range[0]!=beta_schedule[0] or beta_range[-1]!=beta_schedule[-1]):
+                error_msg = "'beta_range' should be set to None, or a value consistent with 'beta_schedule', for 'beta_schedule_type'='custom'."
+                raise ValueError(error_msg)
+            elif min(beta_schedule)<0:
+                error_msg = "'beta_schedule' cannot include negative values."
+                raise ValueError(error_msg)
         else:
-            raise ValueError("Beta schedule type {} not implemented".format(beta_schedule_type))
-
+            if beta_schedule is not None:
+                error_msg = "'beta_schedule' must be set to None for 'beta_schedule_type' not equal to 'custom'"
+                raise ValueError(error_msg)
+            elif num_sweeps is None:
+                num_sweeps = 1000
+                
+            num_betas, rem = divmod(num_sweeps,num_sweeps_per_beta)
+            
+            if rem > 0 or num_betas < 1:
+                error_msg = "'num_sweeps' must be a positive value divisible by 'num_sweeps_per_beta'."
+                raise ValueError(error_msg)
+            
+            if beta_range is None:
+                beta_range = _default_ising_beta_range(bqm.linear, bqm.quadratic)
+            elif len(beta_range) != 2 or min(beta_range) < 0:
+                error_msg = "'beta_range' should be a 2-tuple, or 2 element list of positive numbers. The latter value is the target value."
+                raise ValueError(error_msg)
+            
+            if num_betas == 1:
+                #One sweep in the target model
+                beta_schedule = np.array([beta_range[-1]],dtype=float)
+            else:
+                if beta_schedule_type == "linear":
+                    # interpolate a linear beta schedule
+                    beta_schedule = np.linspace(*beta_range, num=num_betas)
+                elif beta_schedule_type == "geometric":
+                    if min(beta_range) <= 0:
+                        error_msg = "'beta_range' must contain non-zero values for 'beta_schedule_type' = 'geometric'"
+                        raise ValueError(error_msg)
+                    # interpolate a geometric beta schedule
+                    beta_schedule = np.geomspace(*beta_range, num=num_betas)
+                else:
+                    raise ValueError("Beta schedule type {} not implemented".format(beta_schedule_type))
```

Comment:
Input values validation and default values generation above _seem ok_ to me at the first glance.

But to make sure:
- the old behavior is not broken, please add regression tests,
- the new functionality works as expected, please add some tests.